pkg/receive/capnproto_writer.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package receive
+
+import (
+	"context"
+
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+
+	"github.com/thanos-io/thanos/pkg/receive/writecapnp"
+)
+
+type CapNProtoWriterOptions struct {
+	TooFarInFutureTimeWindow int64 // Unit: nanoseconds
+}
+
+type CapNProtoWriter struct {
+	logger    log.Logger
+	multiTSDB TenantStorage
+	opts      *CapNProtoWriterOptions
+}
+
+func NewCapNProtoWriter(logger log.Logger, multiTSDB TenantStorage, opts *CapNProtoWriterOptions) *CapNProtoWriter {
+	if opts == nil {
+		opts = &CapNProtoWriterOptions{}
+	}
+	return &CapNProtoWriter{
+		logger:    logger,
+		multiTSDB: multiTSDB,
+		opts:      opts,
+	}
+}
+
+func (r *CapNProtoWriter) Write(ctx context.Context, tenantID string, wreq *writecapnp.Request) error {
+	tLogger := log.With(r.logger, "tenant", tenantID)
+
+	s, err := r.multiTSDB.TenantAppendable(tenantID)
+	if err != nil {
+		return errors.Wrap(err, "get tenant appendable")
+	}
+
+	app, err := s.Appender(ctx)
+	if err == tsdb.ErrNotReady {
+		return err
+	}
+	if err != nil {
+		return errors.Wrap(err, "get appender")
+	}
+	getRef := app.(storage.GetRef)
+	var (
+		ref          storage.SeriesRef
+		errorTracker = &writeErrorTracker{}
+	)
+	app = &ReceiveAppender{
+		tLogger:        tLogger,
+		tooFarInFuture: r.opts.TooFarInFutureTimeWindow,
+		Appender:       app,
+	}
+
+	var series writecapnp.Series
+	for wreq.Next() {
+		wreq.At(&series)
+
+		var lset labels.Labels
+		// Check if the TSDB has cached reference for those labels.
+		ref, lset = getRef.GetRef(series.Labels, series.Labels.Hash())
```

Comment:
The other writer has a bunch of validations of over labels, like if we have duplicate labels or labels are missing `__name__`, do we need that here?

I dont see those validations on the marshal part of the captnproto, that is why I ask.

Reply:
We were sorting labels in the unmarshal method which I though removed the need for validation.
But since there's more than just to validate, I removed the sort and we do the validation in the writer now.

Long term we should do this in the router to keep CPU usage in the ingestor lower.